packages/drivers/odsp-driver/src/fetchSnapshot.ts

```diff
@@ -58,7 +64,83 @@ export async function fetchSnapshot(
     );
 }
 
-export async function fetchLatestSnapshotCore(
+export async function fetchSnapshotWithRedeem(
+    odspResolvedUrl: IOdspResolvedUrl,
+    storageTokenFetcher: (options: TokenFetchOptions, name: string) => Promise<string | null>,
+    snapshotOptions: ISnapshotOptions | undefined,
+    logger: ITelemetryLogger,
+    snapshotDownloader: (url: string, fetchOptions: {[index: string]: any}) => Promise<IOdspResponse<IOdspSnapshot>>,
+    putInCache: (valueWithEpoch: IVersionedValueWithEpoch) => Promise<void>,
+    removeEntries: () => Promise<void>,
+): Promise<ISnapshotCacheValue> {
+    try {
+        try {
+            const odspSnapshot = await fetchLatestSnapshotCore(
+                odspResolvedUrl,
+                storageTokenFetcher,
+                snapshotOptions,
+                logger,
+                snapshotDownloader,
+                putInCache,
+            );
+            return odspSnapshot;
+        } catch(error) {
+            if (isRedeemSharingLinkError(odspResolvedUrl, error)) {
+                logger.sendErrorEvent({
+                    eventName: "RedeemFallback",
+                    errorType: error.errorType,
+                });
+                await redeemSharingLink(odspResolvedUrl, storageTokenFetcher, logger);
+                const odspResolvedUrlWithoutShareLink: IOdspResolvedUrl =
+                    { ...odspResolvedUrl, sharingLinkToRedeem: undefined };
+                const odspSnapshot = await fetchLatestSnapshotCore(
+                    odspResolvedUrlWithoutShareLink,
+                    storageTokenFetcher,
+                    snapshotOptions,
+                    logger,
+                    snapshotDownloader,
+                    putInCache,
+                );
+                return odspSnapshot;
+            }
+            throw error;
+        }
+    } catch(error) {
+        // Clear the cache on 401/403/404 on snapshot fetch from network because this means either the user doesn't
+        // have permissions for the file or it was deleted. So, if we do not clear cache, we will continue fetching
+        // snapshot from cache in the future.
+        if (typeof error === "object" && error !== null && error.errorType === DriverErrorType.authorizationError
+            || error.errorType === DriverErrorType.fileNotFoundOrAccessDeniedError) {
+            await removeEntries();
+        }
+        throw error;
+    }
+}
+
+async function redeemSharingLink(
+    odspResolvedUrl: IOdspResolvedUrl,
+    storageTokenFetcher: (options: TokenFetchOptions, name: string) => Promise<string | null>,
+    logger: ITelemetryLogger,
+) {
+    return PerformanceEvent.timedExecAsync(
+        logger,
+        {
+            eventName: "RedeemShareLink",
+        },
+        async () => {
```

Comment:
nit - you do not need away of {} here, can do
async () => getWithRetryForTokenRefresh(.....)